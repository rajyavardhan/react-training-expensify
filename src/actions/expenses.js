import uuid from 'uuid';

//Action generators
//ADD_EXPENSE
export const addExpense  = ({createdAt = 0, description = '', notes = '', amount = 0} = {}) => ({
	type: 'ADD_EXPENSE',
	expense: {
		id: uuid(),
		notes,
		amount,
		description,
		createdAt
	}
});

//REMOVE EXPENSE

export const removeExpense = ({removeExpenseId = -1} = {}) => ({
	type: 'REMOVE_EXPENSE',
	removeExpenseId
});

//Edit expense
export  const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});