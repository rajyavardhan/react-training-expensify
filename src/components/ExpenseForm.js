import React, {Component} from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

const now = moment();

export default class ExpenseForm extends Component {
	state = {
		description: this.props.expense.description || '',
		notes: this.props.expense.notes || '',
		amount:this.props.expense.amount||  '',
		createdAt:  moment(this.props.expense.createdAt) || moment(),
		calenderFocused: false,
		errorState: false
	};
	onDescriptionChange = (e) => {
		const description = e.target.value;
		this.setState(()=>({
			description
		}));
	};
	onNoteChange = (e) => {
		const notes = e.target.value;
		this.setState(() => ({
			notes
		}));
	};
	onAmountChange = (e) => {
		const amount = e.target.value;
		if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(()=> ({amount}));
		}
		
	};
	onDateChanged = (createdAt) => {
		if(createdAt){
			this.setState(()=>({
				createdAt
			}));
		}
	};

	onFocusedChanged = ({focused}) => {
		this.setState(()=>({
			calenderFocused:focused
		}));
	};

	onSubmit = (e) => {
		e.preventDefault();
		if(!this.state.description || !this.state.amount) {
			this.setState(()=>({
				errorState: true
			}));
		} else {
			this.setState(()=>({
				errorState: false
			}));
			this.props.onSubmit({
				description: this.state.description,
				createdAt:this.state.createdAt.valueOf(),
				amount: parseFloat(this.state.amount, 10)*100,
				notes: this.state.notes
			});
			//console.log('submitted')
		}
	};
	render() {
		return (
			<div>
				{this.state.errorState && <p>Please submit both Description and amount</p>}
				<form onSubmit = {this.onSubmit}>
				<input 
					type = 'text'
					placeholder = 'Description'
					autoFocus 
					value = {this.state.description}
					onChange = {this.onDescriptionChange}
					/>
					<br />
				<input
					type = 'text'
					placeholder = 'amount'
					value = {this.state.amount}
					onChange = {this.onAmountChange}
					/>
				<br />
				<SingleDatePicker 
					date = {this.state.createdAt}
					onDateChange = {this.onDateChanged}
					focused = {this.state.calenderFocused}
					onFocusChange = {this.onFocusedChanged}
					isOutsideRange = {()=>false}
					numberOfMonths = {1}
				/>
				<textarea
					placeholder = 'Add the notes'
					value = {this.state.notes}
					onChange = {this.onNoteChange}
				>
				</textarea>
				<br />
				<button> Add Expense </button>
				</form>
			</div>
		)
	}
}
ExpenseForm.defaultProps = {
	expense: {
		description: '',
		notes:  '',
		amount: '',
		createdAt: moment()
	}
}