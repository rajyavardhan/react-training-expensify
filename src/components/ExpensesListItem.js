import React from 'react';
import {Link} from 'react-router-dom';
const ExpenseListItem = (props) => (
	<div>
		<Link to = {`/edit/${props.id}`}>
			<h2> {props.description} </h2>
		</Link>
		<span> {props.amount} </span>
		<span> {props.createdAt} </span>
	</div>
);

export default ExpenseListItem;
		// <button
		// 	onClick = {() => {props.history.push('/edit/' +  props.id)}}
		// >
		// 	Edit
		// </button>