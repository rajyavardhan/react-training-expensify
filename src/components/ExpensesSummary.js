import React from 'react';
import { connect } from 'react-redux';
import getExpensesTotal from '../selectors/expenses-total';
import getVisibleExpenses from '../selectors/visibleExpenses';
import numeral from 'numeraljs'

export const ExpensesSummary = (props) => (
	<div>
		{`Viewing ${props.expenses.length} totalling ${numeral(getExpensesTotal(props.expenses)).format('$0.00')}`}
	</div>	
);

const mapStateToProps = (state) => ({
	expenses: getVisibleExpenses(state.expenses, state.filters)
}); 
export default connect(mapStateToProps)(ExpensesSummary);