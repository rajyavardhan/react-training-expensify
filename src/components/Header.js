import React from 'react';
import {NavLink} from 'react-router-dom';
const Header = () => (
	<header>
		<h1>Expensify</h1>
		<NavLink to = "/" activeClassName = "is-active" exact = {true}>HOME</NavLink>
		<NavLink to = "/create" activeClassName = "is-active">CREATE</NavLink>
		<NavLink to = "/HELP" activeClassName = "is-active">HELP</NavLink>
	</header>
);
export default Header;