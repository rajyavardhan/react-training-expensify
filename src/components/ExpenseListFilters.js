import React from 'react';
import {connect} from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import {DateRangePicker} from 'react-dates';

export class ExpenseListFilters extends React.Component {
	state = {
		calenderFocused: null
	}
	onDatesChange = ({startDate, endDate}) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	}
	onFocusChange = (calenderFocused) => {
		this.setState(()=>({
			calenderFocused
		}));
	}
	onInputChange = (e) => {
		this.props.setTextFilter(e.target.value);
	}
	onSelectChange = (e) => {
		if(e.target.value == 'date') {
			this.props.sortByDate();
		} else if(e.target.value == 'amount') {
			this.props.sortByAmount();
		}					
	}
	render() {
		return (
			<div>
				<input type = 'text' value = {this.props.filters.text}
					onChange = {this.onInputChange}
				/>
				<select value = {this.props.filters.sortBy} onChange = {this.onSelectChange}>
					<option value = 'date'>Date</option>
					<option value = 'amount'>Amount</option>
				</select>
				<DateRangePicker 
				 startDate = {this.props.filters.startDate}
				 endDate = {this.props.filters.endDate}
				 onDatesChange = {this.onDatesChange}
				 focusedInput = {this.state.calenderFocused}
				 onFocusChange = {this.onFocusChange}
				 showClearDates = {true}
				 numberOfMonths = {1}
				 isOutsideRange = {() => false}
				/>
			</div>
		)
	}
}

const mapStateToProps = (state) => ({
	filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate)),
	setTextFilter: (value) => dispatch(setTextFilter(value)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount())
});
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);