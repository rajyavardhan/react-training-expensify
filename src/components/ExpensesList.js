import React from 'react';
import {connect} from 'react-redux';
import ExpensesListItem from './ExpensesListItem';
import getVisibileExpenses from '../selectors/visibleExpenses';

export const ExpensesList = (props) => (
	<div>
		<h1>Expense List</h1>
		{props.expenses.length == 0? (<p> There are no Items</p>):(props.expenses.map((expense) => (<ExpensesListItem history = {props.history} key = {expense.id} {...expense}/>))

		)}
		
	</div>
);
const mapStateToProps = (state) => {
	return {
		expenses: getVisibileExpenses(state.expenses, state.filters)
	}
};
export default connect(mapStateToProps)(ExpensesList);
