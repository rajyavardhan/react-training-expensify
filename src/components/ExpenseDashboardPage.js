import React from 'react';
import ExpenseList from './ExpensesList';
import ExpenseListFilters from './ExpenseListFilters';
const ExpenseDashboardPage = (props) => (
	<div>
		<ExpenseListFilters />
		<ExpenseList history = {props.history} store = {props.store}/>
	</div>
);
export default ExpenseDashboardPage;