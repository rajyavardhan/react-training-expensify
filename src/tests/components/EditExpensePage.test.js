import React from 'react';
import {shallow} from 'enzyme';
import {EditExpensePage} from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';
let removeExpense, editExpense, id, history, wrapper;

beforeEach(() => {
	id = {params: {id:2}};
	history = {push: jest.fn()};
	editExpense = jest.fn();
	removeExpense = jest.fn();
	wrapper = shallow(<EditExpensePage 
					editExpense = {editExpense} 
					removeExpense = {removeExpense} 
					expense = {expenses[0]}
					match = {id}
					history = {history}/>);
});

test(`Should render EditExpensePage correctly`, () => {
	expect(wrapper).toMatchSnapshot();
});

test(`Should call removeExpense correctly`, () => {
	wrapper.find('button').simulate('click', 1);
	expect(removeExpense).toHaveBeenCalledWith(2);
});

test(`Should call onSubmit correctly`, () => {
	wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
	expect(editExpense).toHaveBeenCalledWith(2,expenses[1]);
});

test(`Should call history correctly`, () => {	
	wrapper.find('button').simulate('click', 1);
	expect(history.push).toHaveBeenCalledWith('/');
});