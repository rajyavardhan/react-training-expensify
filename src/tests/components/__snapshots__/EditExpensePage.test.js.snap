// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": "19500",
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "notes": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
